spring-kafka/src/test/java/org/springframework/kafka/retrytopic/AsyncMonoRetryTopicScenarioTests.java

```diff
@@ -0,0 +1,1241 @@
+				shortFailedMsg1,
+				shortFailedMsg2,
+				shortFailedMsg3,
+				shortFailedMsg1,
+				shortFailedMsg2,
+				shortFailedMsg3,
+				shortFailedMsg1,
+				shortFailedMsg2,
+				shortFailedMsg3,
+				shortFailedMsg1,
+				shortFailedMsg2,
+				shortFailedMsg3,
+				shortFailedMsg1,
+				shortFailedMsg2,
+				shortFailedMsg3
+		};
+		String[] expectedReceivedTopics = {
+				TEST_TOPIC0,
+				TEST_TOPIC0,
+				TEST_TOPIC0,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic
+		};
+		String[] expectedDltMsgs = {
+				shortFailedMsg1,
+				shortFailedMsg2,
+				shortFailedMsg3
+		};
+
+		// When
+		kafkaTemplate.send(TEST_TOPIC0, shortFailedMsg1);
+		kafkaTemplate.send(TEST_TOPIC0, shortFailedMsg2);
+		kafkaTemplate.send(TEST_TOPIC0, shortFailedMsg3);
+
+		// Then
+		assertThat(awaitLatch(latchContainer.countDownLatch0)).isTrue();
+		assertThat(awaitLatch(latchContainer.dltCountdownLatch0)).isTrue();
+
+		waitAWhile(executor, 10000);
+
+		assertThat(destinationTopic.getDestinationName()).isEqualTo(TEST_TOPIC0 + "-retry");
+
+		assertThat(zeroTopicListener.receivedMsgs).containsExactlyInAnyOrder(expectedReceivedMsgs);
+		assertThat(zeroTopicListener.receivedTopics).containsExactlyInAnyOrder(expectedReceivedTopics);
+		assertThat(latchContainer.extraCountDownLatch0.getCount()).isEqualTo(1000 - CountDownLatchContainer.COUNT0);
+
+		assertThat(myCustomDltProcessor0.receivedMsg).containsExactlyInAnyOrder(expectedDltMsgs);
+		assertThat(latchContainer.extraDltCountdownLatch0.getCount()).isEqualTo(1000 - CountDownLatchContainer.DLT_COUNT0);
+	}
+
+	@Test
+	void firstShortFailAndLastLongSuccessRetryTest(
+			@Autowired TestTopicListener1 testTopicListener1,
+			@Autowired MyCustomDltProcessor myCustomDltProcessor1,
+			@Autowired ThreadPoolTaskExecutor executor) {
+		// Given
+		String longSuccessMsg = "3";
+		String shortFailedMsg = "1";
+		DestinationTopic destinationTopic = topicContainer.getNextDestinationTopicFor("1-topicId", TEST_TOPIC1);
+
+		String expectedRetryTopic = TEST_TOPIC1 + "-retry";
+		String[] expectedReceivedMsgs = {
+				shortFailedMsg,
+				longSuccessMsg,
+				shortFailedMsg,
+				shortFailedMsg,
+				shortFailedMsg,
+				shortFailedMsg
+		};
+
+		String[] expectedReceivedTopics = {
+				TEST_TOPIC1,
+				TEST_TOPIC1,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic
+		};
+
+		String[] expectedDltMsgs = {
+				shortFailedMsg
+		};
+
+		// When
+		kafkaTemplate.send(TEST_TOPIC1, shortFailedMsg);
+		kafkaTemplate.send(TEST_TOPIC1, longSuccessMsg);
+
+		// Then
+		assertThat(awaitLatch(latchContainer.countDownLatch1)).isTrue();
+		assertThat(awaitLatch(latchContainer.dltCountdownLatch1)).isTrue();
+
+		waitAWhile(executor, 10000);
```

Comment:
10 seconds is too long to sleep for nothing in the unit test.
Now image we have 1000 of tests in the project and each of it does such a sleep for nothing.
Plus take into account how really many threads we can use on CI/CD to make some of them to sleep so long.
Cannot we look into the way to condition it via some async barrier instead?

Do we really need so many tests to cover this functionality?
Do they really all have to be so complex?

Reply:
The reason why i use `waitAWhile(executor, 10000)` is for a more accurate test.
I was concerned about the following scenario.

This is the expected case.
1. Initialize the latch (latch = 3), set the retry count to 3, and initialize the DLT latch to 3.
2. Consumer Failed -> latch 2
3. Consumer Retry 1 -> latch 1
4. Consumer Retry 2 -> latch 0
5. Consumer DLT -> DLT latch 0

What about this case? 
1. Initialize the latch (latch = 3), set the retry count to 3, and initialize the DLT latch to 3.
2. Consumer Failed -> latch 2
3. Consumer Retry 1 -> latch 1
4. Consumer Retry 2 -> latch 0
5. Consumer Retry 3 (Unexpected retry) -> latch 0  (latch should raise complain, but not)
7.  Consumer Retry 4 (Unexpected retry) -> latch 0 (latch should raise complain, but not)
8.  Consumer Retry 5 (Unexpected retry) -> latch 0 (latch should raise complain, but not)
9.  Consumer DLT -> DLT latch 0

What about this case? 
1. Initialize the latch (latch = 3), set the retry count to 3, and initialize the DLT latch to 3.
2. Consumer Failed -> latch 2
3. Consumer Retry 1 -> latch 1
4. Consumer Retry 2 -> latch 0
5. Consumer DLT -> DLT latch 0
6. Consumer DLT (Unexpected dlt) -> DLT latch 0 (latch should raise complain, but not)


When the latch overflows, it doesn't raise any complaints. 
In such case, we might consider the test a success even though it actually fails.

However, this could be due to my misunderstanding about the async barrier, as well. 
As you mentioned, I think 10 sec sleep is too long for a unit test case. 

We can remove `waitAWhile(exeuctor, 10000)` or decrease waiting time to 1000 ~ 3000. (it also too long, i guess. but too short value is meaningless i think)

What do you think?

Reply:
I believe we just need one latch for the number of retries. And some holders (or mocks) for interim actions if we want to be sure that something happens in between. The problem with your tests that they are too complex to digest and understand the logic we expect. In the end it is going to be some burden to maintain such a complexity in the tests from now on. Especially when we have so non-reliable timing barriers and for so long period. 

In general the point of tests is to call real code and verify it outcome , but not more code which does not make thing easier to reflect on.
I also cannot tell what and how to make simpler in these tests since I’m not get well with this retry topic feature. 

Maybe @sobychacko will judge us with his fresh perspective on these new tests?

Reply:
Yes, Please give me you guys opinions! 

FYI, 
If we remove `waitAWhile(executor, 10000)` from these test cases, It doesn't affect the test result. 
I added `waitAWhile(executor, 10000)` due to the concern mentioned before, but in reality, that concern did not occur. 😅

Reply:
@artembilan @chickenchickenlove I will take a look today.

Reply:
@chickenchickenlove Do you see the test ever wait for 10 seconds while running it? We have other tests where 10 seconds might be a default wait time, but they rarely wait for that long to get the latch released. If this test indeed takes 10 seconds, then that might be an issue.

Reply:
@sobychacko , @artembilan Thanks for you guys comments.
I fixed it! 

At the first, I imagined a special edge case. 
If the async retry feature might affect the consumer offset (for example, if offset 2 success, offset 1 fails), 
I was worried about a case where a message could be re-consumed infinitely due to seeking the wrong offset. 

However, after reconsidering, we don't need `waitAWhile(...)` to check this edge case. 
```java
protected final List<String> receivedMsgs = new ArrayList<>();
private final List<String> receivedTopics = new ArrayList<>();
```
because in these test cases, the listener writes all the messages and topics it receives.
So, we can check this edge case by asserting that both `receivedMsgs` and `receivedTopics` are equal  to the `expectedValue`.